### ISSUE TITLE
capture: We may need to also capture ShaderReflection/ProgramLayout as well

### ISSUE THREAD

**Issue by kaizhangNV** [shader-slang/slang#4624]
`ShaderReflection`/`ProgramLayout` is where to get or create `slang::TypeReflection` object, without capturing this interface, we cannot replay those reflection API.
Replayer cannot allocate object itself, it can only create object by calling the captured slang APIs.

**Comment by kaizhangNV** [comment]
+ @csyonghe to confirm the correctness of this issue.

**Comment by csyonghe** [comment]
You are right, but we should be able to omit most of the functions around ProgramLayout and TypeReflection, and instead just make sure we can replay the api to obtain the TypeReflection* objects, namely `IComponentType::getLayout` and `spReflection_findTypeByName`.

**Comment by csyonghe** [comment]
Oh, we can also obtain a type from a VariableReflection* and FunctionReflection*, so any methods that can return these objects also neeed to be captured.

**Comment by csyonghe** [comment]
The biggest problem here is that these are C API instead of COM methods, and we need a different mechanism to capture these C API calls. A easy way is to insert a prelude logic at all the C APIs we cared about, which checks a global variable to see if we are capturing, and if so log the call into the recording stream. We shouldn't need any pointer wrapping or translation logic.

**Comment by kaizhangNV** [comment]
But don't we have the C++ class wrapper for those C APIs?
Like TypeLayoutReflection, VariableReflection, VariableLayoutReflection, etc..

We can add capture logic for all of those classes.

I think we discourage people to use those C APIs.

**Comment by csyonghe** [comment]
We do not. The C based reflection API is probably going to stay since refactoring that into COM interfaces would be a lot of work.

**Comment by csyonghe** [comment]
The C++ wrapper is a header-only code that simply calls into the C functions, and I don't think it would be easier to capture those c++ calls comparing to just capture the c calls.
